File: app/frontend/src/types/order.ts
/**
 * Delivery mode is the crux of the checkout flow:
 *  - "self"  → ship to the buyer.
 *  - "gift"  → ship to a recipient, optionally with a gift message and hidden prices.
 */
export type DeliveryMode = 'self' | 'gift';

export interface Address {
  fullName: string;
  line1: string;
  line2?: string;
  city: string;
  state: string;
  postalCode: string;
  country: string;
  phone?: string;
}

export interface GiftDetails {
  recipientName: string;
  message?: string;
  /** Hide prices on the packing slip when sending as a gift. */
  hidePrices: boolean;
}

export interface CheckoutContact {
  email: string;
}

/** Payload posted to the backend `POST /orders` endpoint. */
export interface CreateOrderRequest {
  contact: CheckoutContact;
  deliveryMode: DeliveryMode;
  shippingAddress: Address;
  gift?: GiftDetails;
  items: Array<{
    productId: string;
    size: string;
    quantity: number;
  }>;
}

/** A priced line on a confirmed order (snapshot taken at purchase time). */
export interface OrderLineItem {
  productId: string;
  name: string;
  size: string;
  quantity: number;
  unitPriceCents: number;
  lineTotalCents: number;
}

/** Mirrors the order-service `OrderRead` response. */
export interface Order {
  id: string;
  status: 'pending' | 'confirmed' | 'shipped' | 'delivered' | 'cancelled';
  deliveryMode: DeliveryMode;
  email?: string;
  currency: string;
  subtotalCents?: number;
  shippingCents?: number;
  totalCents: number;
  items: OrderLineItem[];
  gift?: GiftDetails;
  createdAt: string;
}

File: app/backend/order-service/app/schemas/order.py
"""Order API schemas — the wire contract for POST /api/orders.

Mirrors the frontend's `src/types/order.ts`. camelCase on the wire, snake_case in
Python. Input is validated strictly so bad orders are rejected with a clear 422
before any catalogue calls or writes happen.
"""

from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, EmailStr, Field
from pydantic.alias_generators import to_camel

DeliveryMode = Literal["self", "gift"]


class CamelModel(BaseModel):
    model_config = ConfigDict(
        alias_generator=to_camel,
        populate_by_name=True,
        from_attributes=True,
    )


# ── Input ────────────────────────────────────────────────────────────────────
class ContactIn(CamelModel):
    email: EmailStr


class AddressIn(CamelModel):
    full_name: str = Field(..., min_length=1, max_length=160)
    line1: str = Field(..., min_length=1, max_length=200)
    line2: str | None = Field(default=None, max_length=200)
    city: str = Field(..., min_length=1, max_length=100)
    state: str = Field(..., min_length=1, max_length=100)
    postal_code: str = Field(..., min_length=1, max_length=20)
    country: str = Field(..., min_length=1, max_length=100)
    phone: str | None = Field(default=None, max_length=40)


class GiftIn(CamelModel):
    recipient_name: str = Field(..., min_length=1, max_length=160)
    message: str | None = Field(default=None, max_length=400)
    hide_prices: bool = True


class OrderItemIn(CamelModel):
    product_id: str = Field(..., min_length=1, max_length=64)
    size: str = Field(..., min_length=1, max_length=16)
    quantity: int = Field(..., ge=1, le=99)


class CreateOrderRequest(CamelModel):
    contact: ContactIn
    delivery_mode: DeliveryMode
    shipping_address: AddressIn
    gift: GiftIn | None = None
    items: list[OrderItemIn] = Field(..., min_length=1)


# ── Output ───────────────────────────────────────────────────────────────────
class OrderItemRead(CamelModel):
    product_id: str
    name: str
    size: str
    quantity: int
    unit_price_cents: int
    line_total_cents: int


class OrderRead(CamelModel):
    id: str
    status: str
    delivery_mode: DeliveryMode
    email: str
    currency: str
    subtotal_cents: int
    shipping_cents: int
    total_cents: int
    items: list[OrderItemRead]
    gift: GiftIn | None = None
    created_at: datetime
